src/mapblock.cpp

```diff
@@ -381,38 +383,46 @@ void MapBlock::serialize(std::ostream &os, u8 version, bool disk, int compressio
 	/*
 		Bulk node data
 	*/
+	u8 content_width = 2;
+	u8 params_width = 2;
+	writeU8(os, content_width);
+	writeU8(os, params_width);
 	NameIdMapping nimap;
+	SharedBuffer<u8> buf;
 	if(disk)
 	{
 		MapNode *tmp_nodes = new MapNode[nodecount];
 		for(u32 i=0; i<nodecount; i++)
 			tmp_nodes[i] = data[i];
 		getBlockNodeIdMapping(&nimap, tmp_nodes, m_gamedef->ndef());
 
-		u8 content_width = 2;
-		u8 params_width = 2;
-		writeU8(os, content_width);
-		writeU8(os, params_width);
-		MapNode::serializeBulk(os, version, tmp_nodes, nodecount,
-				content_width, params_width, compression_level);
+		buf = MapNode::serializeBulk(version, tmp_nodes, nodecount,
+				content_width, params_width);
 		delete[] tmp_nodes;
 	}
 	else
 	{
-		u8 content_width = 2;
-		u8 params_width = 2;
-		writeU8(os, content_width);
-		writeU8(os, params_width);
-		MapNode::serializeBulk(os, version, data, nodecount,
-				content_width, params_width, compression_level);
+		buf = MapNode::serializeBulk(version, data, nodecount,
+				content_width, params_width);
+	}
+	if(version >= 29) {
+		os.write(reinterpret_cast<char*>(*buf), buf.getSize());
+	} else {
+		// prior to 29 node data was compressed individually
+		compress(buf, os, version, compression_level);
 	}
 
 	/*
 		Node metadata
 	*/
-	std::ostringstream oss(std::ios_base::binary);
-	m_node_metadata.serialize(oss, version, disk);
-	compressZlib(oss.str(), os, compression_level);
+	if (version >= 29) {
+		m_node_metadata.serialize(os, version, disk);
+	} else {
+		// use oss from above to avoid allocating another stream object
```

Comment:
That would be fine. Stream objects are initialized very quickly and this sort of optimization does not seem to be necessary here. It also makes the purpose of `oss` very obscure, because they're for two separate things.

Reply:
Also this compression step is not task of `MapBlock` and should be moved to `NodeMetadata::serialize`.

Reply:
I found creation of string streams to be very slow.

Reply:
https://github.com/minetest/minetest/pull/10788#issuecomment-756967513

Hmm apparently. I wonder what's the cause for this :thinking:. Recycling these streams makes sense, but without proper naming and call structure (like suggested for MapNode in another comment) this results in code that's more difficult to understand and work with in the future.

Reply:
Apparently it has to do locale handling upon creation.

Reply:
That sounds like it shouldn't apply to binary stringstreams, should it? :thinking:

Reply:
There's no way I could find speeding up stringstream creation.